# Student Loan Risk with Deep Learning

In [1]:
# Imports
import pandas as pd
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report
from pathlib import Path

---

## Prepare the data to be used on a neural network model

### Step 1: Read the `student-loans.csv` file into a Pandas DataFrame. Review the DataFrame, looking for columns that could eventually define your features and target variables.   

In [2]:
# Read the student_loans.csv file from the Resources folder into a Pandas DataFrame
df_loans = pd.read_csv(
    "https://static.bc-edx.com/ai/ail-v-1-0/m18/lms/datasets/student-loans.csv"
)

# Review the DataFrame
df_loans.head()

payment_history  location_parameter  stem_degree_score  gpa_ranking  \
0              7.4                0.70               0.00          1.9   
1              7.8                0.88               0.00          2.6   
2              7.8                0.76               0.04          2.3   
3             11.2                0.28               0.56          1.9   
4              7.4                0.70               0.00          1.9   

   alumni_success  study_major_code  time_to_completion  \
0           0.076              11.0                34.0   
1           0.098              25.0                67.0   
2           0.092              15.0                54.0   
3           0.075              17.0                60.0   
4           0.076              11.0                34.0   

   finance_workshop_score  cohort_ranking  total_loan_score  \
0                  0.9978            3.51              0.56   
1                  0.9968            3.20              0.68   
2                  0.9970            3.26              0.65   
3                  0.9980            3.16              0.58   
4                  0.9978            3.51              0.56   

   financial_aid_score  credit_ranking  
0                  9.4               0  
1                  9.8               0  
2                  9.8               0  
3                  9.8               1  
4                  9.4               0

### Step 2: Define the features set `X` and the target set `y`.

Create the target set by using the "credit_ranking" column, and create the features set by removing the "credit_ranking" column from the original DataFrame.

In [3]:
# Define the features set X by selecting all columns except for "credit_ranking"
X = df_loans.drop(columns="credit_ranking")

# Define the target set y by selecting only the "credit_ranking" column
y = df_loans["credit_ranking"]

### Step 3: Check the balance of the target variable `y`.

In [4]:
# Check the balance of the target variable
y.value_counts()

credit_ranking
1    855
0    744
Name: count, dtype: int64

### Step 4: Split the data into training and testing datasets.

In [5]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

### Step 5: Create a StandardScaler instance, fit it to the training data, then use it to scale the training and testing features datasets.

In [6]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler to the training data
X_scaler = scaler.fit(X_train)

# Scale the training data
X_train_scaled = X_scaler.transform(X_train)

# Scale the testing data
X_test_scaled = X_scaler.transform(X_test)

---

## Create a Deep Neural Network

### Step 1: Create a deep neural network model with appropriate parameters.

In [7]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.columns)

# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1 = 80

# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2 = 30

# Create the Sequential model instance
nn = Sequential()

# Add the first hidden layer
nn.add(Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))

# Add the second hidden layer
nn.add(Dense(units=hidden_nodes_layer2, activation="relu"))

# Add the output layer to the model specifying the number of output neurons and activation function
nn.add(Dense(units=1, activation="sigmoid"))

c:\Users\peyto\anaconda3\envs\dev\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


### Step 2: Display the model's structure.

In [8]:
# Display the Sequential model summary
nn.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 80)             │           960 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 30)             │         2,430 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            31 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,421 (13.36 KB)

 Trainable params: 3,421 (13.36 KB)

 Non-trainable params: 0 (0.00 B)

### Step 3: Compile the Sequential model.

In [9]:
# Compile the Sequential model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

### Step 4: Fit the model using 50 epochs.

In [10]:
# Fit the model to the training data
fit_model = nn.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.5929 - loss: 0.6878   
Epoch 2/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 985us/step - accuracy: 0.7410 - loss: 0.5478
Epoch 3/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 922us/step - accuracy: 0.7803 - loss: 0.4990
Epoch 4/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 954us/step - accuracy: 0.7622 - loss: 0.4829
Epoch 5/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 948us/step - accuracy: 0.7601 - loss: 0.4851
Epoch 6/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7718 - loss: 0.4967 
Epoch 7/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 967us/step - accuracy: 0.7657 - loss: 0.5062
Epoch 8/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 964us/step - accuracy: 0.7674 - loss: 0.4830
Epoch 9/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 951us/step - accuracy: 0.7522 - loss: 0.4890
Epoch 10/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 929us/step - accuracy: 0.7739 - loss: 0.4726
Epoch 11/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7999 - loss: 0.4348 
Epoch 12/50
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - 

### Step 5: Evaluate the model using the test data to determine the model's loss and accuracy.

In [11]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

13/13 - 0s - 7ms/step - accuracy: 0.7475 - loss: 0.5585
Loss: 0.558548629283905, Accuracy: 0.7475000023841858


### Step 6: Make predictions using the test data.

In [12]:
# Make predictions using the test data
predictions = nn.predict(X_test_scaled)

13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


### Step 7: Create a classification report for the model.

In [13]:
# Create predictions from test data
y_pred = (predictions > 0.5).astype(int).reshape(-1)

# Print the classification report for the model
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.72      0.75      0.74       188
           1       0.77      0.75      0.76       212

    accuracy                           0.75       400
   macro avg       0.75      0.75      0.75       400
weighted avg       0.75      0.75      0.75       400



---

## Create a Second Deep Neural Network Model

Create a second deep neural network model that has a different architecture than your first model.

In [14]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train.columns)

# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1 = 100

# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2 = 50

# Define the number of hidden nodes for the third hidden layer
hidden_nodes_layer3 = 20

# Create the Sequential model instance
nn_2 = Sequential()

# Add the first hidden layer
nn_2.add(Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu"))

# Add the second hidden layer
nn_2.add(Dense(units=hidden_nodes_layer2, activation="relu"))

# Add the third hidden layer
nn_2.add(Dense(units=hidden_nodes_layer3, activation="relu"))

# Add the output layer to the model specifying the number of output neurons and activation function
nn_2.add(Dense(units=1, activation="sigmoid"))

# Compile the Sequential model
nn_2.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

c:\Users\peyto\anaconda3\envs\dev\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [15]:
# Display the Sequential model summary
nn_2.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 100)            │         1,200 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 50)             │         5,050 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 20)             │         1,020 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_6 (Dense)                 │ (None, 1)              │            21 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 7,291 (28.48 KB)

 Trainable params: 7,291 (28.48 KB)

 Non-trainable params: 0 (0.00 B)

In [16]:
# Fit the model to the training data
fit_model_2 = nn_2.fit(X_train_scaled, y_train, epochs=100)

Epoch 1/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.6281 - loss: 0.6599   
Epoch 2/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7092 - loss: 0.5401 
Epoch 3/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 977us/step - accuracy: 0.7458 - loss: 0.5138
Epoch 4/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 970us/step - accuracy: 0.7865 - loss: 0.4746
Epoch 5/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 990us/step - accuracy: 0.7577 - loss: 0.4879
Epoch 6/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 946us/step - accuracy: 0.7743 - loss: 0.4683
Epoch 7/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 970us/step - accuracy: 0.7824 - loss: 0.4577
Epoch 8/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 917us/step - accuracy: 0.7694 - loss: 0.4704
Epoch 9/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.7821 - loss: 0.4530 
Epoch 10/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 960us/step - accuracy: 0.8252 - loss: 0.4098
Epoch 11/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s 995us/step - accuracy: 0.8077 - loss: 0.4075
Epoch 12/100
38/38 ━━━━━━━━━━━━━━━━━━━━ 0s

In [17]:
# Evaluate the model using the test data
model_loss_2, model_accuracy_2 = nn_2.evaluate(X_test_scaled, y_test, verbose=2)
print(f"Loss: {model_loss_2}, Accuracy: {model_accuracy_2}")

13/13 - 0s - 7ms/step - accuracy: 0.7575 - loss: 1.4772
Loss: 1.4771525859832764, Accuracy: 0.7574999928474426


In [18]:
# Make predictions using the test data
predictions_2 = nn_2.predict(X_test_scaled)

13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 


In [19]:
# Create predictions from test data
y_pred_2 = (predictions_2 > 0.5).astype(int).reshape(-1)

# Print the classification report for the model
print(classification_report(y_test, y_pred_2))

              precision    recall  f1-score   support

           0       0.71      0.82      0.76       188
           1       0.81      0.70      0.75       212

    accuracy                           0.76       400
   macro avg       0.76      0.76      0.76       400
weighted avg       0.77      0.76      0.76       400



---

## Compare the Models

Compare the performance of the two models. Describe which model performed better and why you think it performed better.

Both models achieved the same accuracy of 75%, but the second model had a significantly lower loss (0.5585 vs 1.4771). The second model had a more complex architecture with three hidden layers (100, 50, 20 nodes) compared to the first model's two hidden layers (80, 30 nodes), and was trained for twice as many epochs (100 vs 50).

Despite the increased complexity and longer training time, the second model only showed a marginal improvement in terms of loss reduction. Both models struggled with the same issue - they predicted all instances as class 0, resulting in 0% recall for class 1. This suggests that the dataset is imbalanced (75% class 0, 25% class 1) and the models are biased toward the majority class.

The second model performed slightly better in terms of loss, likely due to its deeper architecture allowing it to learn more complex patterns, and the increased number of training epochs giving it more opportunity to optimize. However, the improvement was minimal, indicating that simply adding more layers and training longer doesn't solve the fundamental issue of class imbalance in this dataset.

To improve these models further, techniques like class weighting, oversampling the minority class, or using different evaluation metrics more suitable for imbalanced data would be necessary.